Add index hints

stevemcleod opened the pull request:
Please review this code. It is a simple implementation of index hints, based on MySQL's implementation.

Syntax:
SELECT * FROM TEST USE INDEX (idx1, idx2)

With table alias:
SELECT * FROM TEST AS t USE INDEX (idx1, idx2)

This could break existing code that uses a table alias called "USE".

stevemcleod committed:
Added parsing of table hints

stevemcleod committed:
Implemented USE INDEX hints

stevemcleod committed:
Added docs, tidied up code for submission

stevemcleod committed:
Added docs, tidied up code for submission

lukaseder commented:
Interesting. I'd strongly suggest making the added syntax also T-SQL compatible:
https://msdn.microsoft.com/en-us/library/ms187373.aspx

svladykin commented:
Since `USE` is not a keyword right now we may break some existing code. I don't like this.

May be it makes sense to listen to @lukaseder and use T-SQL `WITH` clause? It is already a keyword.

lukaseder commented:
> Since USE is not a keyword right now we may break some existing code. I don't like this.

You could have the parser look ahead and only treat `USE` as a keyword if followed by `INDEX`. In that case, you won't have any ambiguity for users wanting to use `USE` as a table alias (without `AS` keyword).

stevemcleod commented:
There was previous discussion on the design of index hints here:
https://groups.google.com/forum/#!topic/h2-database/dHwbBitzXlY

stevemcleod commented:
Good feedback, @lukaseder  and @svladykin .

I'll do nothing more on this for a day or two, to get more feedback if possible.

grandinj commented:
I'd prefer not to invent new syntax, and I'd prefer to follow other open
source databases where possible, which is why I like this syntax. The look
ahead idea is a good one
On Fri, 06 Jan 2017 at 14:11, Steve McLeod <notifications@github.com> wrote:

> Good feedback, @lukaseder <https://github.com/lukaseder> and @svladykin
> <https://github.com/svladykin> .
>
> I'll do nothing more on this for a day or two, to get more feedback if
> possible.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/h2database/h2database/pull/425#issuecomment-270892123>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAwl2TnAA7QWgrFfu9pXblIYQShUymKOks5rPi9mgaJpZM4Lcpb0>
> .
>

stevemcleod commented:
Travis reports has this error (and I got it locally too):

```
org.h2.table.IndexHints.useIndexList = []
org.h2.test.unit.TestClearReferences Tomcat may clear the field above when reloading the web app

```
I'm afraid I don't understand what this indicates, nor what I need to do to fix it... can anyone help?

thomasmueller commented:
> I'm afraid I don't understand what this indicates

The problem is that Tomcat (and other containers) needs to unload classes,
and to do that they use reflection to set static fields to null (of classes
that they think are no longer needed). The is quite a hack (they use
setAccessible(true) and so on). It can lead to problems, and in the past it
did lead to problems for H2. So H2 uses workarounds. TestClearReferences
uses the same logic as Tomcat in clearing fields, so if this test fails
then there is trouble. There are two main solutions:

* Ensure the code works even if the fields are cleared
* Ensure that Tomcat doesn't try to reset the fields

So, either:

* avoid using a static NONE field (public static final IndexHints NONE =
new IndexHints()), and use null instead.
* deal with the case that useIndexList can be null (even if it's final;
final is unfortunately not really final).

In this case, I'm not sure what is the best solution, I guess both options
are fine (and there may be more). Just ensure that this test doesn't fail.

Regards,
Thomas



On Fri, Jan 6, 2017 at 1:14 PM, Steve McLeod <notifications@github.com>
wrote:

> Travis has this error (and I got it locally too):
>
> org.h2.table.IndexHints.useIndexList = []
> org.h2.test.unit.TestClearReferences Tomcat may clear the field above when reloading the web app
>
>
> I'm afraid I don't understand what this indicates, nor what I need to do
> to fix it... can anyone help?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/h2database/h2database/pull/425#issuecomment-270892613>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA46GZa8ZU-jce4YQbGM6ik05bwkB7T0ks5rPjA-gaJpZM4Lcpb0>
> .
>

stevemcleod committed:
Removed used of final static field that would cause problems with Tomcat class unloading

stevemcleod commented:
@thomasmueller 

Thanks for that explanation. I've changed my code to work correctly with Tomcat class unloading.

stevemcleod committed:
Removed used of final static field that would cause problems with Tomcat class unloading

stevemcleod committed:
A table with no index hints now has indexHints == null, instead of an employ instance of IndexHints. For the common case of no index hints, this will prevent some unnecessary object creation.

stevemcleod committed:
Index Hints now is backward compatible with using the "USE" as a table alias

stevemcleod committed:
Added TestIndexHints to list of all tests

stevemcleod commented:
I've now updated the USE INDEX hints functionality to be backward compatible with "USE" as a table alias.

I've also changed to use null instead of an empty instance of IndexHints for the common case where no hints are specified.

svladykin commented:
A couple of comments:

1. Do we really need `IndexHints.useOnlySpecifiedIndexes` field?
2. `TableFilter.getPlanSQL` should contain the respective `USE INDEX` part.

stevemcleod commented:
@svladykin writes

> Do we really need IndexHints.useOnlySpecifiedIndexes field?

Probably not, but I was thinking ahead to adding IGNORE INDEX syntax too, in which case I'd need some way of indicating whether this IndexHints object is USE or IGNORE.

> TableFilter.getPlanSQL should contain the respective USE INDEX part.

Good point.

stevemcleod committed:
Added index hints to text generated by TableFilter.getPlanSQL();

stevemcleod commented:
I've now removed the boolean IndexHints.useOnlySpecifiedIndexes field. If I add IGNORE INDEX hints in the future, I can add a boolean at that time to indicate the type of hints this is.

TableFilter.getPlanSQL now includes the respective USE INDEX part. To ensure the outputted order of the index names matches the order used in the original SQL query, I've changed from HashSet to LinkedHashSet for storing the index names.

svladykin commented:
I think we should have a test for `getPlanSQL` with `USE INDEX` checking the correct order of the indexes in the query plan. 

Another thing is a matter of taste of course, but it looks weird to me that in `IndexHints` we have a field `indexList` which is really a `Set`, but not a `List`. It would be a good idea to name it `allowedIndexes` and rename all the accessors and variables accordingly.

stevemcleod committed:
Improved field names (a field called indexList was actually a set);
Added test for ensuring getPlanSQL lists index hints in correct order

stevemcleod commented:
Good call. I started with a list, changed it a set, and didn't rename the field and methods. It's now called `allowedIndexes`.

> I think we should have a test for getPlanSQL with USE INDEX checking the correct order of the indexes in the query plan.

I've added one simple test for this, but it is not thorough. I'm relying on Java's LinkedHashSet working as documented. However I've explicitly changed the allowedIndexes field to always be a LinkedHashSet rather than a more general Set.

svladykin commented:
In the test it would be enough to check that order is correct in two cases: `use index(idx1, idx2)` and `use index(idx2, idx1)`. Of course I believe that `LinkedHashSet` works correctly but I want to make sure that some occasional future change in `Parser` will not fill it in wrong order.

Otherwise looks good to me!

stevemcleod committed:
Improved unit test for making sure getPlanSQL renders index hints list in correct order

stevemcleod commented:
> In the test it would be enough to check that order is correct in two cases: use index(idx1, idx2) and use index(idx2, idx1).

Sounds reasonable. I've added this to the relevant test.

svladykin commented:
Looks good! Thanks for your work!
